Store delegation reward account balances in the database

rvl opened the pull request:
### Issue Number

ADP-302 / #1750 / #1759

### Overview

Turns out querying reward account balances can be slow.
This is not good if it's done when listing/getting wallets.
Better to do these queries in a separate thread and cache the results in the database.

- [x] Add new database table for the latest known wallet reward account balance.
- [x] Use the database when getting the reward account balance.
- [x] Add a network layer method to watch the node tip.
- [x] Update the account balance in each wallet when the node tip changes (asynchronously).
- [x] Store reward account balance in db as it arrives.
- [x] Adapt Jörmungandr network layer

rvl committed:
Store delegation reward account balances in the database

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -845,11 +848,29 @@ deleteWallet ctx wid = db & \DBLayer{..} -> do
+    _ <- withExceptT ErrFetchRewardsNoSuchWallet
+        . mapExceptT atomically
+        . withNoSuchWallet wid
+        $ readCheckpoint pk
```
Minor note but If this is for triggering a "NoSuchWallet" error, I don't think it is necessary. Since each wallet is stored in an independent database, the only way to get a `ctx` that has a `DBLayer` is to open a corresponding database for that wallet.

rvl replied:
OK I have adjusted a few things to take into account that the WorkerCtx implies an existing wallet.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -881,6 +902,40 @@ fetchRewardBalance ctx wid = db & \DBLayer{..} -> do
+    -> WalletId
+    -> IO ()
+manageRewardBalance ctx wid = db & \DBLayer{..} ->
+    unsafeRunExceptT $ -- TODO: map error type, not that it really matters
```
I think this is _fine_ though, provided that the thread running this method would catch and log properly the error. If it's using the "Worker" machinery, then it'll be done "for free".

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -881,6 +902,40 @@ fetchRewardBalance ctx wid = db & \DBLayer{..} -> do
+            Left _err ->
+                -- Occasionaly failing to query is generally not fatal. It will
+                -- just update the balance next time the tip changes.
+                pure ()
```
Not fatal, yes. Shouldn't it deserve a small warning though?

rvl replied:
It's logged a few lines above.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2010,3 +2075,6 @@ instance HasSeverityAnnotation WalletLog where
         MsgPaymentCoinSelectionAdjusted _ -> Debug
+        MsgRewardBalanceQuery _ -> Debug
+        MsgRewardBalanceResult (Right _) -> Debug
+        MsgRewardBalanceResult (Left _) -> Notice
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -580,17 +586,13 @@ mkShelleyWallet
-        if False
-            then W.fetchRewardBalance @_ @s @t @k wrk wid
-            else pure $ Quantity 0
+        W.fetchRewardBalance @_ @s @k wrk wid
```
Would this call actually _fails_ if the reward balance isn't yet available?

rvl replied:
I added a commend

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1613,8 +1617,9 @@ registerWorker ctx wid =
-                W.restoreWallet @(WorkerCtx ctx) @s @t ctx' wid
+            concurrently_
+                (unsafeRunExceptT $ W.restoreWallet @(WorkerCtx ctx) @s @t ctx' wid)
+                (coworker @(WorkerCtx ctx) ctx' wid)
```
It'll be necessary to do the same thing for `initWorker` too. Otherwise, newly created wallet won't have any reward monitoring in place.

rvl replied:
OK fixed, thanks

KtorZ commented on lib/core/src/Cardano/Wallet/DB.hs:
```diff
@@ -184,6 +184,24 @@ data DBLayer m s k = forall stm. (MonadIO stm, MonadFail stm) => DBLayer
+        -- ^ Get the reward account balance.
+        --
+        -- Returns zero if the wallet isn't found or if wallet hasn't delegated
+        -- stake.
```
Okay :+1: , this answers my previous question.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -673,6 +674,23 @@ newDBLayer trace defaultFieldValues mDatabaseFile = do
+                Nothing -> pure $ Left $ ErrNoSuchWallet wid
+                Just _  -> Right <$> repsert
+                    (DelegationRewardKey wid)
+                    (DelegationReward wid amt)
```
Thinking about rollbacks for a second: it is possible to rollback to a _previous epoch_. Rewards between two forks might actually be slightly different. I don't think having the reward balance _slightly_ off for a short period of time is much of an issue, because the result will _eventually_ become correct. 

Thoughts?

rvl replied:
Yes - I think "eventually correct" is the best we that we can do, because it's not feasible to synchronise wallet applyBlock updates with the LocalStateQuery result messages.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -669,6 +687,51 @@ handleMuxError tr onResourceVanished = pure . errorType >=> \case
+stateTVar :: MonadSTMTx m => TVar_ m a -> (a -> (b, a)) -> m b
+stateTVar var f = do
+   s <- readTVar var
+   let (a, s') = f s -- since we destructure this, we are strict in f
```
"destructure", that's the sign you've been doing too much JavaScript indeed :smile:

rvl replied:
Actually I just copy&pasted this function from the stm library... which is maybe also a sign that I've been doing too much JavaScript.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -669,6 +687,51 @@ handleMuxError tr onResourceVanished = pure . errorType >=> \case
+
+setupWatchers
+    :: Tracer IO (WatcherLog e)
+    -> TQueue IO v
```
There's already a lof of concurrent mutable variables down there, since the function only needs an accessor on the tip, I'd actually change this for just a `IO v` and not even mention the `TQueue` here.

rvl replied:
OK good idea

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -669,6 +687,51 @@ handleMuxError tr onResourceVanished = pure . errorType >=> \case
+            done <- atomically $ newEmptyTMVar
+            let finish = atomically . putTMVar done
+            i <- atomically $ stateTVar watchersVar $ \(nextId, watchers) ->
+                (nextId, (nextId + 1, ((nextId, cb, finish):watchers)))
```
:+1:

KtorZ replied:
Alternatively, `register` could take an explicit `name` for the watcher.

rvl committed:
Add an explanation to Shelley.Launch.withPassiveNode

rvl committed:
Add shelley_qa and shelley_testnet to $CARDANO_NODE_CONFIGS

rvl committed:
Fix nightly push of all-tests-pass branch

KtorZ committed:
use 'Chan' instead of 'TQueue' for node's tip, and simplify the watcher logic

Chans have the nice property to be 'duplicable'. So, we can rely on this instead of re-implementing our own fan-out logic. I believe the original error that made worker crashes could have been a race-condition where the watcher would exit unexpectedly but I wasn't quite clearly able to identify the cause. Using 'Chan' looks definitely simpler.

I've already removed the possibility for workers to exit with a failure. This wasn't actually used and added on top of the complexity. If that's needed, we can work it out a bit later but for now, a simpler API makes it easier to debug.

Integration scenarios are no longer 'blowing up' unexpectedly, yet, the 'reward accumulate and stop' scenario still does not pass.

KtorZ committed:
remove unused genesis and node config parameters

rvl committed:
Make BracketTracer aware of exceptions

rvl committed:
More logging around the NetworkLayer.watchNodeTip callback

rvl committed:
Add ToObject instance to BracketLog

rvl committed:
Use bracketTracer in sqlite runQuery

rvl committed:
Fix worker thread exiting on non-shelley wallets

rvl committed:
Update assertion in NETWORK_PARAMS test

rvl committed:
Update haddock in Registry

Anviking committed:
Enable parts of "Rewards accumulate" test

KtorZ committed:
pass coworker as argument for wallets creating from pub keys

These wallet can be either Byron or Shelley wallets, so they don't necessarily have rewards capability.

rvl committed:
More comments about watchNodeTip in byron and jormungandr

KtorZ committed:
disable integration tests which seems to make a false assumption

Our intuition may be wrong here, and we need to investigate this.

KtorZ committed:
turn off assertions which require emptying a reward account.
